src/relations/db.py

```diff
@@ -0,0 +1,259 @@
+                    "version": self.charm.postgresql.get_postgresql_version(),
+                    "user": user,
+                    "password": password,
+                    "database": database,
+                }
+                databag.update(updates)
+        except (
+            PostgreSQLCreateDatabaseError,
+            PostgreSQLCreateUserError,
+            PostgreSQLGetPostgreSQLVersionError,
+        ):
+            self.charm.unit.status = BlockedStatus(
+                f"Failed to initialize {self.relation_name} relation"
+            )
+            return
+
+    def _on_relation_departed(self, event: RelationDepartedEvent) -> None:
+        """Handle the departure of legacy db and db-admin relations.
+
+        Remove unit name from allowed_units key.
+        """
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        if event.departing_unit.app == self.charm.app:
+            # Just run for departing of remote units.
+            return
+
+        departing_unit = event.departing_unit.name
+        local_unit_data = event.relation.data[self.charm.unit]
+        local_app_data = event.relation.data[self.charm.app]
+
+        current_allowed_units = local_unit_data.get("allowed_units", "")
+
+        logger.debug(f"Removing unit {departing_unit} from allowed_units")
+        local_app_data["allowed_units"] = local_unit_data["allowed_units"] = " ".join(
+            {unit for unit in current_allowed_units.split() if unit != departing_unit}
+        )
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Remove the user created for this relation."""
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        # Delete the user.
+        user = f"relation_id_{event.relation.id}"
+        try:
+            self.charm.postgresql.delete_user(user)
+        except PostgreSQLDeleteUserError:
+            self.charm.unit.status = BlockedStatus(
+                f"Failed to delete user during {self.relation_name} relation broken event"
+            )
+
+    def _get_allowed_subnets(self, relation: Relation) -> str:
+        """Build the list of allowed subnets as in the legacy charm."""
+
+        def _csplit(s) -> Iterable[str]:
+            if s:
+                for b in s.split(","):
+                    b = b.strip()
+                    if b:
+                        yield b
```

Comment:
As this is called (I think...) only once, could this be part of the logic in `_get_allowed_subnets`?